sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -55,100 +55,24 @@ object ConstantFolding extends Rule[LogicalPlan] {
 }
 
 /**
- * Substitutes [[Attribute Attributes]] which can be statically evaluated with their corresponding
- * value in conjunctive [[Expression Expressions]]
+ * Substitutes expressions which can be statically narrowed by constrains.
  * eg.
  * {{{
- *   SELECT * FROM table WHERE i = 5 AND j = i + 3
- *   ==>  SELECT * FROM table WHERE i = 5 AND j = 8
+ *   SELECT * FROM table WHERE i = 5 AND j = i + 3     => SELECT * FROM table WHERE i = 5 AND j = 8
+ *   SELECT * FROM table WHERE i <= 5 AND i = 5        => SELECT * FROM table WHERE i = 5
```

Comment:
I don't think this aims the equal goal. Could you make a separate rule instead?

Reply:
+1

Reply:
Filter reduction should be a separate rule.

Reply:
Thank you all for the feedback. The 2 rules seemed similar and easy to combine to me, but I will not mix them then.

Reply:
I separated the 2 rules. I added some improvement to the `ConstantPropagation' and created a new 'FilterReduction`.

Reply:
I've moved all new logic to `FilterReduction`.